### PR TITLE
git:scripts: move custom git hooks into the 'pre-commit' tool

### PR DESCRIPTION
**Pull request by Baekalfen**
To make pre-commit hooks in Git easier to maintain and configure, we
have moved to the software 'pre-commit' from https://pre-commit.com.

This will allow us to make more advanced setups down the line, by also
utilizing the vast number of community plugins for other languages than
C.

Signed-off-by: Mads Ynddal <m.ynddal@samsung.com>

### CODE REVIEW

**Review by safl — CHANGES_REQUESTED** [review]
Quick feedback, to look at this then some information would be helpful, including:

- [ ] What is ``.sort.cfg`` and why is it needed?
- [ ] What is ``.style.yapf`` and why is it needed?
- [ ] What is ``.pre-commit-config.yaml`` and why is it needed?
- [ ] What is required to use the ``pre-commit`` tool? Note-to-self, +1 for #45 

Some additional comments
- It would be preferable that the clang-format-definitions remain such that tools other than ``pre-commit`` can use them e.g. adding them to editors and being able manually run ``source-format.sh``.
- How about just having ``pre-commit`` invoke ``source-format.sh``?
- Can the various dot-files be placed inside the ``scripts`` folder? That would be preferable.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    args: ["-style={BasedOnStyle: LLVM, AlignConsecutiveBitFields: AcrossEmptyLinesAndComments, AlignEscapedNewlines: Left, AlwaysBreakAfterReturnType: All, BreakBeforeBraces: WebKit, ColumnLimit: 99, ContinuationIndentWidth: 8, IndentGotoLabels: false, IndentWidth: 8, MaxEmptyLinesToKeep: 1, SortIncludes: Never, TabWidth: 8, UseTab: ForContinuationAndIndentation}", "-i"]
+    files: ^(include|lib|examples|tests|tools)\/((?!fioe\.c$).)*$ # Match all .c files, but not ones ending in "fioe.c"
+- repo: https://github.com/pre-commit/mirrors-clang-format
+  rev: v13.0.1
+  hooks:
+  - id: clang-format
+    args: ["-style={BasedOnStyle: LLVM, AlignConsecutiveBitFields: AcrossEmptyLinesAndComments, AlignEscapedNewlines: Left, AlwaysBreakAfterDefinitionReturnType: None, AlwaysBreakAfterReturnType: None, BreakBeforeBraces: WebKit, ColumnLimit: 99, ContinuationIndentWidth: 8, IndentGotoLabels: false, IndentWidth: 8, MaxEmptyLinesToKeep: 1, SortIncludes: Never, TabWidth: 8, UseTab: ForContinuationAndIndentation}", "-i"]
+    files: ^(include|lib|examples|tests|tools)\/.*fioe\.c$ # Match all .c files, that _does_ end in "fioe.c"
+
+- repo: https://github.com/pre-commit/mirrors-yapf
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply .pre-commit-config.yaml:66]: It's a clang-format based Python formatter by Google

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  rev: 'v0.32.0'
+  hooks:
+  - id: yapf
+- repo: https://github.com/pre-commit/mirrors-isort
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply .pre-commit-config.yaml:70]: It's an **i**mport-**sort**er for Python files

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  rev: v5.10.1
+  hooks:
+  - id: isort
+- repo: https://github.com/myint/unify
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply .pre-commit-config.yaml:74]: It unifies the usage of single- and double-quotes in Python

File: .style.yapf [hunk 4]
```diff
@@ -0,0 +1,16 @@
+[style]
```
[the remark this answers is not in the document]
Comment (Baekalfen) [reply .style.yapf:1]: It's the style config for yapf

**Comment by Baekalfen** [comment]
I've changed the PR. Now it only includes the same functionality as the current setup.

> * [ ]  What is `.pre-commit-config.yaml` and why is it needed?
That's the configuration file for pre-commit, and where everything is now centralized. If you want to add new hooks, this is where they go.

> * [ ]  What is required to use the `pre-commit` tool? Note-to-self, +1 for #45
If you can install it through `pip3 install pre-commit`, then you're good to go. Other requirements, are based on the hooks we add.

> * It would be preferable that the clang-format-definitions remain such that tools other than `pre-commit` can use them e.g. adding them to editors and being able manually run `source-format.sh`.
This is a limitation of clang-format. When clang-format v14 is released and is available on common distros, we can use separate files.
Does anyone right now use the format files in their editors?

> * How about just having `pre-commit` invoke `source-format.sh`?
You can call anything from `pre-commit`, so that is an option. But I'm not sure what the benefit would be. `source-format.sh` is mostly plumbing, which is the purpose of `pre-commit` as well. If you want to manually call the formatter on all files, that's also possible by `pre-commit run clang-format --all` for example.

> * Can the various dot-files be placed inside the `scripts` folder? That would be preferable.
That's up to each hook to support that. It's considered standard to place these in the root of the repo.

**Comment by safl** [comment]
Thanks for providing info on pre-commit here on the PR and the offline discussion.
I don't see much value in adding ``pre-commit`` as-is, however I do see the potential value in adding ``pre-commit`` hooks for bunch of things, starting with picking some low-hanging fruit by adding some checks in areas that have no help until it hits the CI:

* Check format of ``.github/*/*.yaml``

And then in concert with https://github.com/OpenMPDK/xNVMe/issues/45 , that is, providing ``pre-commit`` hooks such as:
* Check that all function in the public API have valid doc-definitions
* Check format of commit-messages
* Check source-format

Then going beyond that, then using ``pre-commit`` for tasks which will be more time-consuming, but which can be done quickly locally without too much setup:
* Check that the default build-configuration works, that is, invoke ``meson setup/build``
* Check coding-conventions e.g. variable-declarations in nearest scope, variable-declarations-before logic, possibly via custom-hook and definitions for clang-tidy
* Run core-verification by invoking cijoe and exercising the testplans with an environment for the Memory Backend https://github.com/OpenMPDK/xNVMe/issues/46 / https://github.com/OpenMPDK/xNVMe/pull/38

In other words. It would be great with additional scripts and tools to reduce trivial/early CI rejections, PR ping/pong, and generally add tools assisting the contributor as well as the reviewer. The ``pre-commit`` seems like a great framework for the task as it provides functionality for dispatching **which** hooks to run on **what** and conventions for writing the hooks. A good start for adopting ``pre-commit`` would be to address the areas briefly described above without breaking current functionality and maintaining the ability to ignore ``pre-commit`` via ``GIT_HOOKS_OFF=1`` as it should always be a help and not a nuisance.

**Review by safl — CHANGES_REQUESTED** [review]
See latest comment for general feedback on a path for adoption of ``pre-commit``.

**Comment by Baekalfen** [comment]
> ... starting with picking some low-hanging fruit by adding some checks in areas that have no help until it hits the CI

I've tried added a YAML checker/formatter and readded Python formatting, although without any additional files.

> * Check that all function in the public API have valid doc-definitions

Possible, but requires more work.

> * Check format of commit-messages

Done. I've made a simple regex.

> * Check source-format

pre-commit would replace `source-format.sh`, so I'm not sure what you mean.

> * Check that the default build-configuration works, that is, invoke `meson setup/build`

Apart from validating the syntax, I think that should be up to the CI build system.

> * Check coding-conventions e.g. variable-declarations in nearest scope, variable-declarations-before logic, possibly via custom-hook and definitions for clang-tidy

Possible. https://github.com/pocc/pre-commit-hooks adds a hook for `clang-tidy`.

> * Run core-verification by invoking cijoe and exercising the testplans with an environment for the Memory Backend [Memory-backed pseudo-device #46](https://github.com/OpenMPDK/xNVMe/issues/46) / [Memory backend #38](https://github.com/OpenMPDK/xNVMe/pull/38)

Again, that seems like something for the CI system. A pre-commit would normally be static and not take more than a few seconds.

**Comment by safl** [comment]
This is definitely growing on me :)  Regarding the use of the ``pre-commit`` as a guard shielding the wasteful run of potentially also self-hosted runners makes a lot of sense. Regarding this, currently the ``generate-source-archive`` uses a custom docker-container to shave off as many seconds as possible for archive-generation.
Could the same Alpine+Bash+Git image be used at the ``source-format/pre-commit`` job? Potentially it would shave off 20 seconds. Btw. the dependencies on CI infra which xNVMe uses are here:
https://github.com/refenv/gh-automation
We would either add a second alpine-image, keeping the current as small as possible.
Then adding another with the tools required specifically for the ``source-format-check/pre-commit`` stage.

**Comment by Baekalfen** [comment]
> This is definitely growing on me :) 

🥳

> Could the same Alpine+Bash+Git image be used at the `source-format/pre-commit` job? ... Then adding another with the tools required specifically for the `source-format-check/pre-commit` stage.

Yes, easily. The only base requirements are git and a functional Python env. I picked `python:3.10-bullseye` over `uses: actions/setup-python@v3` as that saves time. I saw now, that Git is already installed in the container, so we can skip that part.

**Comment by safl** [comment]
Additional input on this, I am currently working on moving the ``cijoe-pkg-xnvme`` into ``scripts/cijoe-pkg-xnvme``.
Here ``pre-commit``, like you demonstrated, would make sense on the collection of Python + Bash which makes up a cijoe-package. Here I am thinking of hooks for:
* black (to avoid discussing code-style, just adhere to the opinions of black)
* flake (for catching misc stuff.)
* pylint (for catching misc stuff.)
* mypy (might be useful as it seems like the best option for checking dataclasses)
* shellcheck (for Bash-scripts)
cijoe packages currently run the above as CI/itHUB actions. Would be nice a maintain those checks when moving the cijoe-pkg-xnvme into the xNVMe repository.

When adding all those Python+Bash hooks, will that "infest" a contributors system with package-installation? Or only in case they modify Python/Bash? In other words, would pre-commit trigger the installation of the tools it needs unconditionally, or only "just-in-time"?

**Comment by safl** [comment]
Another thing. The "source_format_check" could just use the git-checkout action, and thus run in parallel with the source-archive-generation.
It can do this, since it, unlike the rest, does not depend on subprojects.

**Comment by Baekalfen** [comment]
> * flake (for catching misc stuff.)

I'll add flake8. I was concerned it might conflict with black, but it looks like it should work.

> * pylint (for catching misc stuff.)

Let's try it out, and see if it conflicts with black or flake8.

> * mypy (might be useful as it seems like the best option for checking dataclasses)
> * shellcheck (for Bash-scripts)

I'll see if I can add these.

> When adding all those Python+Bash hooks, will that "infest" a contributors system with package-installation?

The contributor will have to voluntarily install pre-commit (`pip install pre-commit`) and install the hook (`pre-commit install; pre-commit install --hook-type commit-msg`) for anything to happen. Afterwards, when they commit, pre-commit will install its dependencies in a virtual env, separate from the system.

**Comment by safl** [comment]
@Baekalfen thanks for your effort on this, I have re-worked it here: https://github.com/OpenMPDK/xNVMe/pull/60
Please provide your review, closing this one.
